mutt_socket.c

```diff
@@ -423,11 +418,6 @@ int raw_socket_write (CONNECTION* conn, const char* buf, size_t count)
     mutt_error (_("Error talking to %s (%s)"), conn->account.host,
 		strerror (errno));
     mutt_sleep (2);
-  } else if (errno == EINTR) {
```

Comment:
Here is the beginnings of a test case written in Python. It creates an IMAP server then runs mutt and connects to the the server.  

```python
#!/usr/bin/python3

import threading
import socketserver
import pexpect
import os
from tempfile import NamedTemporaryFile

verbose = True

class MyServer(socketserver.TCPServer):
    allow_reuse_address = True

class SimpleIMAPHandler(socketserver.StreamRequestHandler):
    continuation = None

    def setup(self):
        super().setup()
        self.server.logged = None

    def _send(self, message):
        if verbose:
            print("SENT: %r" % message.strip())
        self.wfile.write(message)

    def _send_line(self, message):
        self._send(message + b'\r\n')

    def _send_textline(self, message):
        self._send_line(message.encode('ASCII'))

    def _send_tagged(self, tag, code, message):
        self._send_textline(' '.join((tag, code, message)))

    def handle(self):
        # Send a welcome message.
        self._send_textline('* OK IMAP4rev1')
        while 1:
            line = b''
            while True:
                part = self.rfile.read(1)
                if part == b'':
                    return
                line += part
                if line.endswith(b'\r\n'):
                    break

            if verbose:
                print('GOT: %r' % line.strip())
            if self.continuation:
                try:
                    self.continuation.send(line)
                except StopIteration:
                    self.continuation = None
                continue
            splitline = line.decode('ASCII').split()
            tag = splitline[0]
            cmd = splitline[1]
            args = splitline[2:]

            if hasattr(self, 'cmd_' + cmd):
                continuation = getattr(self, 'cmd_' + cmd)(tag, args)
                if continuation:
                    self.continuation = continuation
                    next(continuation)
            else:
                self._send_tagged(tag, 'BAD', cmd + ' unknown')

    def cmd_CAPABILITY(self, tag, args):
        self._send_textline('* CAPABILITY IMAP4rev1')
        self._send_tagged(tag, 'OK', 'CAPABILITY completed')

    def cmd_LOGOUT(self, tag, args):
        self.server.logged = None
        self._send_textline('* BYE IMAP4ref1 Server logging out')
        self._send_tagged(tag, 'OK', 'LOGOUT completed')

    def cmd_LOGIN(self, tag, args):
        self.server.logged = args[0]
        self._send_tagged(tag, 'OK', 'LOGIN completed')

    def cmd_LIST(self, tag, args):
        self._send_textline(r'* LIST (\Noselect) "/" ""')
        self._send_tagged(tag, 'OK', 'LIST completed')

    def cmd_SELECT(self, tag, args):
        self._send_textline('* 0 EXISTS')
        self._send_textline('* 0 RECENT')
        self._send_tagged(tag, 'OK', '[READ-WRITE] SELECT completed')

    def cmd_NOOP(self, tag, args):
        self._send_tagged(tag, 'OK', 'NOOP completed')

    def cmd_CLOSE(self, tag, args):
        self._send_tagged(tag, 'OK', 'CLOSE completed')

def write_muttrc(port):
    f = NamedTemporaryFile(delete=False)
    muttrc = f.name
    f.write(b'''
set folder=imap://x:x@localhost:%d/
set mail_check=1
set timeout=2
unset help
unset ssl_starttls
''' % port)
    f.close()
    return muttrc

def read_loop(mutt):
    while True:
        try:
            buf = mutt.read_nonblocking(mutt.maxread, timeout=1)
        except pexpect.TIMEOUT:
            break
        print('from mutt:', repr(buf))

def run_mutt(port, muttrc):
    mbox = 'imap://x:x@localhost:{}'.format(port)
    mutt = pexpect.spawn('mutt', ['-F', muttrc, '-f', mbox], encoding='utf-8')

    read_loop(mutt)

    print('send mutt: imap-fetch-mail')
    mutt.send(':exec imap-fetch-mail\n')
    read_loop(mutt)

    print('send mutt: quit')
    mutt.send('q')
    try:
        read_loop(mutt)
    except pexpect.exceptions.EOF:
        pass

def main():
    server = MyServer(('localhost', 0), SimpleIMAPHandler)
    port = server.server_address[1]

    thread = threading.Thread(target=server.serve_forever)
    thread.daemon = True
    thread.start()

    muttrc = write_muttrc(port)
    run_mutt(port, muttrc)
    os.remove(muttrc)


if __name__ == '__main__':
    main()```